pkg/plugin/windows/pktmon/pktmon_plugin_windows.go

```diff
@@ -111,103 +116,135 @@ func (p *Plugin) RunPktMonServer() error {
 }
 
 func (p *Plugin) Start(ctx context.Context) error {
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
-
 	p.enricher = enricher.Instance()
 	if p.enricher == nil {
 		return ErrNilEnricher
 	}
 
 	go func() {
-		err := p.RunPktMonServer()
+		err := p.RunPktMonServer(ctx)
 		if err != nil {
-			p.l.Error("failed to run pktmon server", zap.Error(err))
+			p.l.Error("pktmon server exited", zap.Error(err))
 		}
+	}()
+
+	err := p.SetupStream()
+	if err != nil {
+		return fmt.Errorf("failed to setup initial pktmon stream: %w", err)
+	}
+
+	eventCh := make(chan *v1.Event, eventChannelSize)
 
-		// if the pktmon server process exits, cancel the context, we need to crash
-		cancel()
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				p.l.Info("pktmon plugin context done")
+				return
+			case ev := <-eventCh:
+				if p.enricher != nil {
+					p.enricher.Write(ev)
+				} else {
+					p.l.Error("enricher is nil when writing event")
+				}
+
+				// Write the event to the external channel.
+				if p.externalChannel != nil {
+					select {
+					case p.externalChannel <- ev:
+					default:
+						// Channel is full, drop the event.
+						// We shouldn't slow down the reader.
+						metrics.LostEventsCounter.WithLabelValues(utils.ExternalChannel, string(Name)).Inc()
+					}
+				}
+			}
+		}
 	}()
 
-	var str observerv1.Observer_GetFlowsClient
+	// run the getflows loop
+	for {
```

Comment:
Here’s the second infinite loop for getting flows. Remove one of the infinite loops or rename GetFlow to represent the new logic?

Reply:
the idea here is that p.GetFlow contains a child context, which is represents the lifetime of the streaming rpc. when that rpc bails due to an rpc error, that child context is canceled, and returns here to the caller. if it is an rpc error (which happens a decent amount on windows), restart the streaming rpc via calling GetFlows again. if it's not a known rpc (status) error, bail out of the plugin and crash

Reply:
Makes sense. I feel like adding a couple comments and renaming GetFlow (e.g. make it plural) might help others who read the code later, but I wouldn’t block the PR on it